pkg/blockbuilder/blockbuilder.go

```diff
@@ -104,11 +166,68 @@ func (b *BlockBuilder) starting(context.Context) (err error) {
 
 func (b *BlockBuilder) stopping(_ error) error {
 	b.kafkaClient.Close()
-
 	return nil
 }
 
-func (b *BlockBuilder) running(ctx context.Context) error {
+// runningPullMode is a service `running` function for pull mode, where we learn
+// about jobs from a block-builder-scheduler. We consume one job at a time.
+func (b *BlockBuilder) runningPullMode(ctx context.Context) error {
+	// Kick off the scheduler's run loop.
+	go b.scheduler.Run(ctx)
+
+	for {
+		if err := ctx.Err(); err != nil {
+			if errors.Is(err, context.Canceled) {
+				return nil
+			}
+			return err
+		}
+
+		key, spec, err := b.scheduler.GetJob(ctx)
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				return nil
+			}
+			level.Error(b.logger).Log("msg", "failed to get job", "err", err)
+			continue
+		}
+
+		if _, err := b.consumeJob(ctx, key, spec); err != nil {
+			level.Error(b.logger).Log("msg", "failed to consume job", "job_id", key.Id, "epoch", key.Epoch, "err", err)
+			continue // ?
+		}
+
+		// TODO: CompleteJob needs to accept the new state returned from consumeJob.
+
+		if err := b.scheduler.CompleteJob(key); err != nil {
+			level.Error(b.logger).Log("msg", "failed to complete job", "job_id", key.Id, "epoch", key.Epoch, "err", err)
+		}
+	}
+}
+
+// consumeJob performs block consumption from Kafka into object storage based on the given job spec.
+func (b *BlockBuilder) consumeJob(ctx context.Context, _ schedulerpb.JobKey, jobSpec schedulerpb.JobSpec) (PartitionState, error) {
+	state := PartitionState{
+		Commit: kadm.Offset{
+			Topic:     jobSpec.Topic,
+			Partition: jobSpec.Partition,
+			At:        jobSpec.StartOffset,
+			Metadata:  "{}", // FIXME. This and other fields need to be plumbed in via the scheduler RPC layer.
+		},
+		CommitRecordTimestamp: jobSpec.CommitRecTs,
+		LastSeenOffset:        jobSpec.LastSeenOffset,
+		LastBlockEnd:          jobSpec.LastBlockEndTs,
+	}
+
+	cycleEndTime := jobSpec.LastBlockEndTs // ???
```

Comment:
This is something the scheduler can send as part of the spec. Also, the current implementation of consumePartition() stops consuming after it has hit cycleEndTime. But we should update it to end consuming when it hits the cycleEndOffset perhaps.

Reply:
At the scheduler side, cycleEndTime can be the time when the job was freshly created.

Reply:
OK, cool. I'd like to address this comment in #10105 where I'm attempting to rectify all of the Kafka administrivia bits.

Reply:
Actually, I went ahead and added it.